### ISSUE TITLE
ImportError when importing two different packages generated with binder that share an overlap

### ISSUE THREAD

**Issue by kliegeois** [RosettaCommons/binder#248]
When loading two different packages (say A and B) for which the interface has been generated with binder and if packages A and B share at least the usage of one class ( `std::fpos<__mbstate_t>` in my case ), I have an error message as follows:

```
ImportError: generic_type: type "fpos___mbstate_t_t" is already registered!
```

The type is defined in a binder generated line:
```
pybind11::class_<std::fpos<__mbstate_t>> cl(M("std"), "fpos___mbstate_t_t", "");
```
both in A and B.

I don't want/can't introduce a dependency of B on A or A on B in this case and I need both A and B to work as a stand-alone interface.
To solve the multiple registration of a type, would it make sense to add an option to prepend a string in the function `string python_class_name(CXXRecordDecl const *C)` that would be library dependent such that we would have:

For A:
```
pybind11::class_<std::fpos<__mbstate_t>> cl(M("std"), "A_fpos___mbstate_t_t", "");
```

For B:
```
pybind11::class_<std::fpos<__mbstate_t>> cl(M("std"), "B_fpos___mbstate_t_t", "");
```

**Comment by kliegeois** [comment]
This is a known pybind11 issue https://github.com/pybind/pybind11/issues/439

**Comment by kliegeois** [comment]
This would probably require a modification of binder though.

**Comment by lyskov** [comment]
@kliegeois have you tried compiling and exporting using described approach? What i want to double check is that how Pybind11 reacts to attempts to register same type under different names.

re idea itself: assuming that proposed solution is works (see my above question), i am open to the idea but this seems a bit "messy". I imaging a few packages when each have its own `std::` bindings... so user will ends up with multiple bindings for the same types. I wonder if cleaner solution is possible here, for instance perhaps it is possible to check if bindings for particular type_id was already registered? And if yes then skip bindings procedure. Have you investigate this approach?

**Comment by kliegeois** [comment]
@lyskov the described approach:
> 
> For A:
> 
> ```
> pybind11::class_<std::fpos<__mbstate_t>> cl(M("std"), "A_fpos___mbstate_t_t", "");
> ```
> 
> For B:
> 
> ```
> pybind11::class_<std::fpos<__mbstate_t>> cl(M("std"), "B_fpos___mbstate_t_t", "");
> ```

does not work.

But, 

For A:

```pybind11::class_<std::fpos<__mbstate_t>> cl(M("std"), "_fpos___mbstate_t_t", "", py::module_local());```

For B:

```pybind11::class_<std::fpos<__mbstate_t>> cl(M("std"), "_fpos___mbstate_t_t", "", py::module_local());```

Does work.

It seems to even work when you parse an instance of a class from one package to the other providing that both of them define that class.

I have a PR ready to solve this issue with extra options. I just cannot create the PR right now.

**Comment by lyskov** [comment]
oh, i am glad to hear that passing `py::module_local()` solve this, - this is much cleaner solution, - great! - looking forward for PR thats add this!

**Comment by lyskov** [comment]
re how to solve this: - i wonder if it will make sense to provide options to trigger local bindings both per-class and per-namespace basis 🤔...

**Comment by kliegeois** [comment]
Currently, I implemented it as a per namespace option with one option to do it for all namespaces. Hopefully I can open the PR today.

**Comment by kliegeois** [comment]
Closed by #251.
